### PR TITLE
Add basic support (via fallbacks) for attachments.

### PR DESCRIPTION
**Pull request by Cadair**
Signed-off-by: Stuart Mumford <stuart@cadair.com> 

This is a mitigation for #36 and also includes improvements to how bot users are represented in matrix.

### CODE REVIEW

**Comment by Cadair** [comment]
So, I think I am going to have to block rich text representations of attachments on #110 because I am going to need proper support for doing different things with the plain and html text of the matrix message.

@Half-Shot this can be reviewed when you have time.

File: lib/SlackEventHandler.js [hunk 1]
```diff
@@ -178,42 +211,37 @@ SlackEventHandler.prototype.handleMessageEvent = function (params) {
...
+                msg.file = file;
+                const content = await this.fetchFileContent(msg.file, token);
+                msg.file._content = content;
+            }
         }
-    } else {
-        result = Promise.resolve();
     }
 
-    return result.then(() => msg)
```
Comment (Half-Shot) [inline lib/SlackEventHandler.js]: Be careful here that the original code is returning the result of `onSlackMessage` to the caller, wheras you are returning `undefined`. I don't know if it's an issue.

File: lib/SlackEventHandler.js [hunk 2]
```diff
@@ -161,7 +177,24 @@ SlackEventHandler.prototype.handleMessageEvent = function (params) {
...
+    // Handle events with attachments like bot messages.
+    if (params.event.type === "message" && params.event.attachments) {
+        log.warn(params.event.attachments);
+        for (var attachment of params.event.attachments) {
+            msg.text = attachment.fallback;
+
+            msg = await this.doChannelUserReplacements(msg, token);
+            await room.onSlackMessage(msg);
+        }
+        if (params.event.text != '') {
```
Comment (Half-Shot) [inline lib/SlackEventHandler.js]: Can't help but wonder if we could invert this
```suggestion
        if (params.event.text == '') {
            return;
        }
        msg.text = params.event.text;
```

File: lib/SlackEventHandler.js [hunk 3]
```diff
@@ -161,7 +177,24 @@ SlackEventHandler.prototype.handleMessageEvent = function (params) {
         channel_id: params.event.channel
     });
 
-    const processMsg = msg.text || msg.subtype === 'message_deleted' || msg.files != undefined || msg.subtype === 'message_changed';
+    // Handle events with attachments like bot messages.
+    if (params.event.type === "message" && params.event.attachments) {
+        log.warn(params.event.attachments);
+        for (var attachment of params.event.attachments) {
+            msg.text = attachment.fallback;
+
```
Comment (Half-Shot) [inline lib/SlackEventHandler.js]: newline here feels a bit odd

File: lib/SlackEventHandler.js [hunk 4]
```diff
@@ -161,7 +177,24 @@ SlackEventHandler.prototype.handleMessageEvent = function (params) {
         channel_id: params.event.channel
     });
 
-    const processMsg = msg.text || msg.subtype === 'message_deleted' || msg.files != undefined || msg.subtype === 'message_changed';
+    // Handle events with attachments like bot messages.
+    if (params.event.type === "message" && params.event.attachments) {
+        log.warn(params.event.attachments);
```
Comment (Half-Shot) [inline lib/SlackEventHandler.js]: looks like a leftover

File: lib/SlackEventHandler.js [hunk 5]
```diff
@@ -178,42 +211,37 @@ SlackEventHandler.prototype.handleMessageEvent = function (params) {
...
         log.warn("no slack token for " + room.getSlackTeamDomain() || room.getSlackChannelId());
 
-        return (processMsg) ? room.onSlackMessage(msg) : Promise.resolve();
+        if (processMsg) {
+            await room.onSlackMessage(msg);
+        }
+        return;
     }
 
     if (!processMsg) {
```
Comment (Half-Shot) [inline lib/SlackEventHandler.js]: Could this be moved just after `const processMsg = ...`?

File: lib/SlackEventHandler.js [hunk 6]
```diff
@@ -178,42 +211,37 @@ SlackEventHandler.prototype.handleMessageEvent = function (params) {
         // just send the message as text.
         log.warn("no slack token for " + room.getSlackTeamDomain() || room.getSlackChannelId());
 
-        return (processMsg) ? room.onSlackMessage(msg) : Promise.resolve();
+        if (processMsg) {
```
Comment (Half-Shot) [inline lib/SlackEventHandler.js]: If you move the `if (!processMsg)`, you can drop this.

File: lib/SlackGhost.js [hunk 7]
```diff
@@ -60,34 +60,49 @@ SlackGhost.prototype.update = function(message, room) {
     ]);
 };
 
-SlackGhost.prototype.updateDisplayname = function(message, room) {
+SlackGhost.prototype.getDisplayName = async function(message, room) {
+    const token = room.getAccessToken();
+    if (message.bot_id) {
+        const bot = await this.lookupBotInfo(message.bot_id, token);
+        if (bot && bot.name) {
```
Comment (Half-Shot) [inline lib/SlackGhost.js]:
```suggestion
        return (bot && bot.name) ? bot.name : message.bot_id
```

File: lib/SlackGhost.js [hunk 8]
```diff
@@ -60,34 +60,49 @@ SlackGhost.prototype.update = function(message, room) {
...
+
+    }
+    return;
 };
 
 SlackGhost.prototype.lookupAvatarUrl = function(user_id, token) {
     return this.lookupUserInfo(user_id, token).then((user) => {
-        if (!user || !user.profile) return;
+        if (!user || !user.profile) {
+            log.debug("Can't read user profile");
```
Comment (Half-Shot) [inline lib/SlackGhost.js]:
```suggestion
            log.debug("Can't read user profile, so cannot lookup avatar url");
```

would be more helpful for debugging

**Comment by karsrhyder** [comment]
+1

**Comment by jbrinksmeier** [comment]
looks like @Cadair is too busy finishing this. Would it be ok if somebody takes over @Half-Shot? The requested changes do not look too complicated and this PR does fix issues as is, see #42

**Comment by Half-Shot** [comment]
Sure, whoever fixes first @jbrinksmeier

**Comment by jbrinksmeier** [comment]
@umitalp could you take over?

**Comment by umitalp** [comment]
Sure, I will do the requested changes and do another PR based on this one? @Half-Shot

**Comment by Cadair** [comment]
I am closing this in favour of #126 thanks @umitalp

**Pull request closed without merging**